napari/components/dims.py

```diff
@@ -100,94 +156,120 @@ def _check_dims(cls, values):
+        axis_labels = values['axis_labels']
+        labels_ndim = len(axis_labels)
+        if labels_ndim < ndim:
             # Append new "default" labels to existing ones
-            if values['axis_labels'] == tuple(
-                map(str, range(len(values['axis_labels'])))
-            ):
-                values['axis_labels'] = tuple(map(str, range(ndim)))
+            if axis_labels == tuple(map(str, range(labels_ndim))):
+                updated['axis_labels'] = tuple(map(str, range(ndim)))
             else:
-                values['axis_labels'] = (
-                    tuple(map(str, range(ndim - len(values['axis_labels']))))
-                    + values['axis_labels']
+                updated['axis_labels'] = (
+                    tuple(map(str, range(ndim - labels_ndim))) + axis_labels
                 )
-        elif len(values['axis_labels']) > ndim:
-            values['axis_labels'] = values['axis_labels'][-ndim:]
+        elif labels_ndim > ndim:
+            updated['axis_labels'] = axis_labels[-ndim:]
 
-        return values
+        return {**values, **updated}
 
     @property
-    def nsteps(self) -> Tuple[int, ...]:
-        """Tuple of int: Number of slider steps for each dimension."""
+    def nsteps(self) -> Tuple[float, ...]:
         return tuple(
-            int((max_val - min_val) / step_size)
-            for min_val, max_val, step_size in self.range
+            # "or 1" ensures degenerate dimension works
+            int((rng.stop - rng.start) / (rng.step or 1))
+            for rng in self.range
         )
 
+    @nsteps.setter
+    def nsteps(self, value):
+        self.range = [
+            (rng.start, rng.stop, (rng.stop - rng.start) / nsteps)
+            for rng, nsteps in zip(self.range, value)
+        ]
+
     @property
-    def point(self) -> Tuple[int, ...]:
-        """Tuple of float: Value of each dimension."""
-        # The point value is computed from the range and current_step
-        point = tuple(
-            min_val + step_size * value
-            for (min_val, max_val, step_size), value in zip(
-                self.range, self.current_step
-            )
+    def current_step(self):
+        return tuple(
+            int(round((point - rng.start) / (rng.step or 1)))
+            for point, rng in zip(self.point, self.range)
+        )
+
+    @current_step.setter
+    def current_step(self, value):
+        self.point = [
+            rng.start + point * rng.step
+            for point, rng in zip(value, self.range)
+        ]
+
+    @property
+    def thickness(self) -> Tuple[float, ...]:
+        return tuple(
+            left + right
+            for left, right in zip(self.margin_left, self.margin_right)
         )
-        return point
+
+    @thickness.setter
+    def thickness(self, value):
+        self.margin_left = self.margin_right = tuple(val / 2 for val in value)
 
     @property
-    def displayed(self) -> Tuple[int, ...]:
+    def displayed(self) -> Tuple[int]:
```

Comment:
Nit: this typing change (and the ones around it) is inconsistent with many of the additions here. Either justify it and make the other changes consistent, or just leave it alone.